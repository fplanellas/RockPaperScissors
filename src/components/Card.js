import React from 'react';
import styled from 'styled-components';

const CardStyled = styled.div`
 
`

const Card = () => {
  return (
    <CardStyled>
       card
    </CardStyled>
  )
}

export default Card
