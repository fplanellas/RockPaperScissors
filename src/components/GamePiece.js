import React from 'react';
import styled from 'styled-components';

const GamePieceStyled = styled.div`
    width: 100px;
    height: 100px;
    border: 5px solid #9b9b9b;
    border-radius: 50%;
    display: flex; 
    justify-content: center;
    align-items: center;
    background: silver;
    cursor: pointer;
`

const GamePiece = ({ name = '', clicked }) => {
 
    const handleClick = () => {
        if (clicked) {
            clicked(name);
        };      
    };

  return (
    <GamePieceStyled onClick={ handleClick }>
      <img src={`./assets/${name}.png`} alt="" width="80px"/>
    </GamePieceStyled>
  )
}

export default GamePiece
