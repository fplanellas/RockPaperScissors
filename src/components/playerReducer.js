
export const playerReducer = (state = [], action) => {

    switch (action.type) {
    case 'add':
        return [...state, action.payload];

    case 'score':
        return state.map(player => {
            if (player.id === action.payload) {

                return {
                    ...player,
                    score: action.score
                };
            } else {
                return player;
            }
        });

    default:
        return state;
    }
};