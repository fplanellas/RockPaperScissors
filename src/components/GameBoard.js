import React, { useContext, useState } from 'react';
import styled from 'styled-components';
import GamePiece from './GamePiece';
import { ScoreContex } from './GameScreen';
import { GameBoardContex } from './GameScreen';
import { GameBoardInit } from './GameScreen';

const GameBoardStyled = styled.div`
  position: relative;
  top: 6em;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  gap: 15px;
  text-align: center;
  .result{
      width: 100%;
  }
  .none {
      display: none;
  }

  button {
      border: 1px solid silver;
      border-radius: 5px;
      background: #fff;
      color: #565656; 
      width: 80%;
      padding: 8px;
      font-weight: bold;
      margin-top: 10px;
      text-transform: uppercase;
  }
`;

const GameBoard = () => {

    const [showResults, setShowResults] = useState(false);
    const { gameB } = useContext(GameBoardContex);
    const { score, setScore } = useContext(ScoreContex);
    const { userPlaying, setUserPlaying } = useContext(GameBoardInit);
    const [choose, setChose] = useState('');
    const [botChoose, setBotChose] = useState('');
    const [results, setResults] = useState('');

    const handlePlayAgain = () => {
        setUserPlaying(false);
        setResults('');
        setShowResults(false);
    };

    const botRandomChoice = (min, max) => {
        return Math.floor(Math.random() * (max - min)) + min;
    };

    const fortune = (choose, botChoose) => {

        if (choose === botChoose) {
            return "It's a tie";
        }

        if (choose === 'paper') {
            if (botChoose === 'rock') {
                return 'You win';
            }
            if (botChoose === 'scissors') {
                return 'You lose';
            }
        }

        if (choose === 'rock') {
            if (botChoose === 'paper') {
                return 'You lose';
            }
            if (botChoose === 'scissors') {
                return 'You win';
            }
        }

        if (choose === 'scissors') {
            if (botChoose === 'paper') {
                return 'You win';
            }
            if (botChoose === 'rock') {
                return 'You lose';
            }
        }
    };

    const elements = [
        'rock',
        'scissors',
        'paper'
    ];

    const launchBotChoose = () => {
        return new Promise((resolve, reject) => {

            let choose;

            const interval = setInterval(() => {
                choose = elements[botRandomChoice(0, 3)];
                setBotChose(choose);
            }, 100);

            setTimeout(() => {
                clearInterval(interval);
                resolve(choose);
                setShowResults(true);
            }, 2000);


        });
    };

    const clicked = async (name) => {
        setShowResults(false);
        setUserPlaying(true);
        setChose(name);
        const bot = await launchBotChoose();
        const results = fortune(name, bot);
        setResults(results);

        if (results === 'You win') {
            setScore(score + 1);
        }
    };

    return (
        <div>
            {
                gameB === false && (
                    <GameBoardStyled userPlaying={userPlaying}>
                        {
                            !userPlaying ? (
                                <>
                                    <GamePiece name="rock" clicked={clicked} />
                                    <GamePiece name="scissors" clicked={clicked} />
                                    <GamePiece name="paper" clicked={clicked} />
                                </>
                            ) : (
                                <>
                                    <div>
                                        <GamePiece name={choose} />
                                        <p>You choose</p>
                                    </div>
                                    <div>
                                        <GamePiece name={botChoose} />
                                        <p>Bot choose</p>
                                    </div>
                                    <div className={`${showResults === false ? 'none' : 'result'}`}>
                                        <h2>{results}</h2>
                                        <button onClick={handlePlayAgain}>
                                            Play again
                                        </button>
                                    </div>
                                </>
                            )
                        }
                    </GameBoardStyled>
                )
            }
        </div>
    );
};

export default GameBoard;
