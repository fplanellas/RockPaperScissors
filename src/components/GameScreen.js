import React, { createContext, useState } from 'react';
import Header from "./Header";
import styled from 'styled-components';
import Wrapper from './Wrapper';
import GameBoard from './GameBoard';

const ScreenStyled = styled.main`
    background: #000;
    padding: 0.5em;
    position: relative;
    height: 500px;
    color:#fff;
  `;

export const ScoreContex = createContext();
export const GameBoardContex = createContext();
export const GameBoardInit = createContext();

const GameScreen = () => {

    const [score, setScore] = useState(0);
    const [gameB, setGameB] = useState(true);
    const [userPlaying, setUserPlaying] = useState(false);

    return (
        <GameBoardInit.Provider value={{
            userPlaying,
            setUserPlaying
        }}>
            <GameBoardContex.Provider value={{
                gameB,
                setGameB
            }}>
                <ScoreContex.Provider value={{
                    score,
                    setScore
                }}>
                    <ScreenStyled>
                        <Wrapper>
                            <Header />
                            <GameBoard />
                        </Wrapper>
                    </ScreenStyled>
                </ScoreContex.Provider>
            </GameBoardContex.Provider>
        </GameBoardInit.Provider>
    );
};

export default GameScreen;
