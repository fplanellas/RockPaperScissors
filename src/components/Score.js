import React, { useContext, useEffect, useReducer, useState } from 'react';
import styled from 'styled-components';
import useForm from '../hooks/useForm';
import { ScoreContex } from './GameScreen';
import { GameBoardContex } from './GameScreen';
import { GameBoardInit } from './GameScreen';
import { playerReducer } from './playerReducer';

const ScoreStyled = styled.div`
   text-align: center;
   padding: 5px 0; 
   color: #fff;
   position: absolute;
   left:0;
   width: 100%;
   top: 0;
   small {
       width:100%;
       text-align: center;
       font-size:15px;
   }
   p {
       margin:auto;
       padding:0
       width:100%;
       font-size: 35px;
       font-weight: bold;
       text-align: center;
   }
   form {
       z-index:200;
        width: 100%;      
       margin-right: 50px;
       display: flex;
        flex-direction: column;
        align-items: center;
        gap: 15px;
   }
   input {
    width:300px;
    background: #212121;
    border: 2px solid rgba(255, 255, 255, .29);
    padding:10px;
    color:#bdbdbd;
    font-size: 18px;
    border-radius: 5px;
    }  
    button {
        width:300px;
        border:2px solid grey;
        border-radius: 5px;
        background: #bdbdbd;
        padding:10px;
        color: #000;
        font-weight: bold;
    }

   h3 {
        margin-top: 100px;
        font-size: 18px;
   }
   .container-score-screen {
    display: flex;
    flex-wrap: wrap;
    align-content: space-between;
    gap: 15px;
   }
   .score-screen {     
       display: flex;
       flex-direction: column;
        align-items: center;
        border: 2px solid rgba(255, 255, 255, .29);
        border-radius: 10px;
        padding: 0.5em 1em;
        margin: auto;
        margin-top: 60px;

   }
   .save-back {
       position: absolute;
       right: 12px;
       top: 14px;
       cursor: pointer;
   }
` 
    const init = () => {
            return JSON.parse(localStorage.getItem('players')) || [];
    };
  
    const Score = () => {
        
        const { score, setScore } = useContext(ScoreContex);
        const [ players, dispatch ] = useReducer(playerReducer, [], init);
        const [ userNaming, setUserNaming ] = useState(false);
        const { gameB, setGameB } = useContext(GameBoardContex);
        const { userPlaying, setUserPlaying } = useContext(GameBoardInit);

        const [ { description }, handleInputChange, reset ] = useForm({
            description: ''
        });

        useEffect( () => {
            localStorage.setItem('players', JSON.stringify( players ))
        }, [players] ) ;

        const handleOut = (player) => {
            const action = {
                type: 'score',
                payload: player.id,
                score: player.score + score
            };

            dispatch( action );
            setUserNaming(false);
            setGameB(true);
            setUserPlaying(false);
        };

        const handleSubmit = (e) => {
            e.preventDefault();
       
            if ( description.trim().length <= 1 ) {
                return;
            }
    
            const newPlayer = {
                id: description,
                name: description,
                score: 0
            };
    
            const action = {
                type: 'add',
                payload: newPlayer
            };
            
            const adding = action.payload;
            const elementAdded = players.find(element => element.id === action.payload.id);
            setScore(score - score);
            
            setUserNaming(true);
            setGameB(false);  

            if (!elementAdded) {
                dispatch( action );               
                window.pan = adding;                           
            }

            if (elementAdded) {              
                window.pan = elementAdded;
            }
            
            reset();
        }
        
    let painting = window.pan;
  
    return (
        <ScoreStyled> 
            <div>
                {
                  userNaming === true ? (
                    <div className='container-score-screen'> 
                        <div className='save-back'> 
                            <img onClick={ () => handleOut(painting)} src="./assets/logout_white.svg"  alt="" width="24px"/> 
                        </div>                                 
                        {
                            painting && (
                                <div className='score-screen'>
                                    <small>{painting.name}</small>
                                    <small>{painting.score + score}</small>
                                </div>                  
                            )
                        }                                            
                    </div>
                  ) : (
                        <>
                            <h3>Create new payer</h3>
                            <form onSubmit={ handleSubmit }>
                                    <input 
                                        type='text'
                                        name='description'
                                        placeholder='player'
                                        autoComplete='off'
                                        value={ description }
                                        onChange={ handleInputChange }
                                    />
                                    <button>
                                        JOIN
                                    </button>
                            </form>
                        </>
                    
                  )
                }
              </div>  
        </ScoreStyled>
    )
}

export default Score
