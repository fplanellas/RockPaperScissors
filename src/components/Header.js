import React from 'react';
import styled from 'styled-components';
import Score from './Score';

const HeaderStyled = styled.div`
    border: 2px solid rgba(255, 255, 255, .29);
    border-radius: 10px;
    padding: 0.5em 1em;
    position:relative;
    h1{
        font-size: 20px;
        margin-top: 6px;
    }
`

const Header = () => {

  return (         
      <HeaderStyled>
          <div>
              <h1>
                  Rock Paper Scissors    
              </h1>

              <Score />

          </div>                           
      </HeaderStyled>    
  )
}

export default Header

