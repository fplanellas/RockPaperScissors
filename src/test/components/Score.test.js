import React from 'react';
import { shallow } from 'enzyme';
import Score from '../../components/Score';
import { ScoreContex } from '../../components/GameScreen';

describe('testing in component Score', () => {

    const score = {
        gamescore: 5
    }
  
    const wrapper = shallow( 
        <ScoreContex.Provider value={{
            score
        }}>
            <Score />
        </ScoreContex.Provider>
    );

    test('should be displayed correctly', () => {
        expect( wrapper ).toMatchSnapshot();
    })
      
})