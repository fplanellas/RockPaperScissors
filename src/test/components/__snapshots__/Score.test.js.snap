// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing in component Score should be displayed correctly 1`] = `<Score />`;
